geo/src/main/java/org/apache/carbondata/geo/scan/expression/PolygonRangeListExpression.java

```diff
@@ -0,0 +1,211 @@
+import org.apache.carbondata.core.scan.filter.resolver.FilterResolverIntf;
+import org.apache.carbondata.core.scan.filter.resolver.RowLevelFilterResolverImpl;
+import org.apache.carbondata.geo.GeoConstants;
+import org.apache.carbondata.geo.GeoHashUtils;
+import org.apache.carbondata.geo.scan.filter.executor.PolygonFilterExecutorImpl;
+
+/**
+ * InPolygonRangeList expression processor. It inputs the InPolygonRangeList string to
+ * the Geo implementation's query method, inputs lists of range of IDs and is to be calculated
+ * the and/or/diff range list to filter. And then, build InExpression with list of all the IDs
+ * present in those list of ranges.
+ */
+@InterfaceAudience.Internal
+public class PolygonRangeListExpression extends UnknownExpression
+    implements ConditionalExpression {
+
+  private String polygonRangeList;
+
+  private String opType;
+
+  private List<Long[]> ranges = new ArrayList<Long[]>();
+
+  private ColumnExpression column;
+
+  private static final ExpressionResult trueExpRes =
+      new ExpressionResult(DataTypes.BOOLEAN, true);
+
+  private static final ExpressionResult falseExpRes =
+      new ExpressionResult(DataTypes.BOOLEAN, false);
+
+  public PolygonRangeListExpression(String polygonRangeList, String opType, String columnName) {
+    this.polygonRangeList = polygonRangeList;
+    this.opType = opType;
+    this.column = new ColumnExpression(columnName, DataTypes.LONG);
+  }
+
+  private void processExpression() {
+    try {
+      // 1. parse the range list string
+      List<String> rangeLists = new ArrayList<>();
+      Pattern pattern = Pattern.compile(GeoConstants.RANGELIST_REG_EXPRESSION);
+      Matcher matcher = pattern.matcher(polygonRangeList);
+      while (matcher.find()) {
+        String matchedStr = matcher.group();
+        rangeLists.add(matchedStr);
+      }
+      // 2. process the range lists
+      if (rangeLists.size() > 0) {
+        List<Long[]> processedRangeList = getRangeListFromString(rangeLists.get(0));
+        for (int i = 1; i < rangeLists.size(); i++) {
+          List<Long[]> tempRangeList = getRangeListFromString(rangeLists.get(i));
+          processedRangeList = GeoHashUtils.processRangeList(
+            processedRangeList, tempRangeList, opType);
+        }
+        ranges = processedRangeList;
+        GeoHashUtils.validateRangeList(ranges);
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void sortRange(List<Long[]> rangeList) {
+    rangeList.sort(new Comparator<Long[]>() {
+      @Override
+      public int compare(Long[] x, Long[] y) {
+        return Long.compare(x[0], y[0]);
+      }
+    });
+  }
+
+  private void combineRange(List<Long[]> rangeList) {
+    if (rangeList.size() > 1) {
+      for (int i = 0, j = i + 1; i < rangeList.size() - 1; i++, j++) {
+        long previousEnd = rangeList.get(i)[1];
+        long nextStart = rangeList.get(j)[0];
+        if (previousEnd + 1 == nextStart) {
```

Comment:
Not handled for overlapping ranges ?

Reply:
Corrected it.  For overlapping case [[2,8], [5,10]] and containing case [[2,8], [5,6]]